x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DocVector.java

```diff
@@ -138,35 +142,57 @@ private void buildShardSegmentDocMapIfMissing() {
             for (int p = 0; p < forwards.length; p++) {
                 forwards[p] = p;
             }
-            new IntroSorter() {
-                int pivot;
-
-                @Override
-                protected void setPivot(int i) {
-                    pivot = finalForwards[i];
-                }
-
-                @Override
-                protected int comparePivot(int j) {
-                    int cmp = Integer.compare(shards.getInt(pivot), shards.getInt(finalForwards[j]));
-                    if (cmp != 0) {
-                        return cmp;
+            if (singleSegment()) {
+                new IntroSorter() {
+                    int pivot;
+
+                    @Override
+                    protected void setPivot(int i) {
+                        pivot = finalForwards[i];
+                    }
+
+                    @Override
+                    protected int comparePivot(int j) {
+                        return Integer.compare(docs.getInt(pivot), docs.getInt(finalForwards[j]));
```

Comment:
This reduces the number of comparisons from 5 to 1.

Reply:
Oh neat. Yeah!